### PR TITLE
WICKET-6862: Add @Autowired annotation to SpringInjector

### PR DESCRIPTION
**Pull request by bhuism**
Besides annotations @Inject and @Springbean also support @Autowired annotation for spring bean injection

### CODE REVIEW

File: wicket-spring/src/main/java/org/apache/wicket/spring/injection/annot/AnnotProxyFieldValueFactory.java [hunk 1]
```diff
@@ -117,18 +118,26 @@ public Object getFieldValue(final Field field, final Object fieldOwner)
 		{
 			SpringBean annot = field.getAnnotation(SpringBean.class);
 
-			String name;
-			boolean required;
+			String name = null;
+			Boolean required = true;
```
Comment (solomax) [inline wicket-spring/src/main/java/org/apache/wicket/spring/injection/annot/AnnotProxyFieldValueFactory.java:122]: I would declare this one as `boolean` it seems to be not nullable :)

**Review by bitstorm — APPROVED** [review]

**Review by svenmeier — CHANGES_REQUESTED** [review]
@Autowired is Spring's own injection annotation.

Since Spring will not actually do the injection on Wicket components, I don't see the need to add support for @Autowired here.

File: wicket-spring/src/main/java/org/apache/wicket/spring/injection/annot/AnnotProxyFieldValueFactory.java (hunk 1 above)
Comment (bhuism) [reply wicket-spring/src/main/java/org/apache/wicket/spring/injection/annot/AnnotProxyFieldValueFactory.java:122]: Good idea, thanks

**Comment by bhuism** [comment]
@svenmeier I agree, it is the spring (defacto) standard injection annotation, and does not _need_ to be added. The rationale is for consistency. In normal spring beans one can (obviously) not use wickets @SpringBean. The other option @Inject is not the spring defacto standard, I think new wicket users (who know spring) will find it intuitive (or even expect) to be able to use @Autowired

ps: I would love to see that Spring does the injection, buts thats another story

**Comment by svenmeier** [comment]
This change needs to be discussed. Please open a Jira issue so other core devs are aware of this.

**Comment by bhuism** [comment]
@svenmeier done, please see: https://issues.apache.org/jira/browse/WICKET-6862

**Comment by martin-g** [comment]
I agree with @svenmeier - by using `@SpringBean` one knows who is responsible to do the injection - Wicket.
If we start supporting `Autowired` then soon we will have to take care for `@Qualifier` too...
If Spring's BeanFactory had a method like `getBean(Field)` then maybe this would have been cleaner. Just pass the field to Spring and let it deal with all possible annotations it supports.

Actually `SpringBean` is an artefact from the past. The idea was everyone (Spring included) to move to javax.inject APIs but this never happened.

**Comment by bhuism** [comment]
Hmm, I hear 'we' want to migrate away from @SpringBean (a thing from the past, and unknown to non wicket users) and away from @Inject because it was never adopted. Can this PR be seen as a step in that direction? toward using spring to do the actual injection? am trying to make wicket better, for widest audience possible. And won't be offended if it doesn't get merged.

**Comment by martin-g** [comment]
We cannot drop `@SpringBean` in a minor version. The earliest would be Wicket v 10.0.0. We can deprecate it though!
But I'd prefer to use `@javax.inject.Inject` instead of `@Autowired`. I doubt that Spring people will ever do it but they can also decide to drop `@Autowired` in favor of `@Inject`

`@Inject` is already supported by wicket-spring, wicket-guice and wicketstuff-jee.

**Comment by bhuism** [comment]
I'll doubt that too, It is a _spring_-injector, because of that I though following spring semantics (`@Inject` and `@Autowired`) would be the way to go. Would you or @svenmeier consider supporting `@Autowired` when spring does the injection?

**Comment by martin-g** [comment]
> when spring does the injection

What do you mean ?

At the moment Wicket delegates to Spring/Guice to find the respective bean and then Wicket just sets the Field value.
What do you want to change in this area ?

**Comment by bhuism** [comment]
Yes, I mean just the setting of the fields part, maybe we can leverage spring (more) in doing the actual injection, a quick search in the spring api gave me:

[`AutowireCapableBeanFactory.autowire(...)`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/config/AutowireCapableBeanFactory.html#autowireBeanProperties-java.lang.Object-int-boolean-)

I think @svenmeier also meant this in https://github.com/apache/wicket/pull/461#pullrequestreview-573137771

This way maybe also annotations like `@Value` can be supported in the future, or other spring bean features.

**Comment by martin-g** [comment]
This looks promising!

**Comment by svenmeier** [comment]
>Would you or @svenmeier consider supporting @Autowired when spring does the injection?

Yes. I remember trying with AutowireCapableBeanFactory a decade ago, but if I recall correctly I didn't find a solution to interject the Wicket proxies.

**Comment by martin-g** [comment]
> Wicket proxies

Right! Now I remember this problem too - Wicket needs to inject a Serializable proxy to be able to store the page. Spring returns the bean as is - it could be Serializable by chance but most of the service classes are not.

**Comment by bhuism** [comment]
Ok, I'll close it. I did expect the _spring_-injector to follow _spring_ semantics as much as possible. I had a different idea with the functionality.

@martin-g is it not the case that beans are not serialized with the page (sorry for not checking) i.e. nulled before serialization, and re-injected after deserialization? if the beans _are_ serialized, why would you want the (stateless) beans to be part of the page serialisation state? I am curious.

**Comment by martin-g** [comment]
Wicket does not keep a hard reference to the Spring/Guice/JEE beans.
Wicket injects a Serializable proxy that lazy loads the bean on first usage. The lazy loading is request scoped, i.e. the beans are transient and the proxy re-loads them for each http request.

**Comment by bhuism** [comment]
a, thanks @martin-g I thought as much

**Pull request closed without merging**